File: arrays-and-strings/string_compression.py
# 1.6: String Compression

# Runtime: O(n) - Space: O(n)
def string_compression(string: str) -> str:
    compressed = []
    curr = string[0]
    count = 0

    for s in string:
        if s != curr:
            compressed.append(curr)
            compressed.append(str(count))

            curr = s
            count = 1
        else:
            count += 1

        if len(compressed) > len(string):
            return string

    compressed.append(curr)
    compressed.append(str(count))

    return "".join(compressed)


print(string_compression("aabcccccaaa"))  # a2b1c5a3

File: 01-arrays-and-strings/javascript/stringCompression.js
// 1.6: String Compression

// Runtime: O(n) - Space: O(n)
function stringCompression(str) {
  const compressed = []
  let curr = str[0]
  let count = 0

  for (const char of str) {
    if (char !== curr) {
      compressed.push(`${curr}${count}`)
      curr = char
      count = 1
    } else {
      count++
    }

    if (compressed.length > str.length) return str
  }

  compressed.push(`${curr}${count}`)
  return compressed.join("")
}

console.log(stringCompression("aabcccccaaa"))  // a2b1c5a3
